_posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md

```diff
@@ -0,0 +1,79 @@
+
+If you are enqueing lots of jobs, you could be using [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing), this is hard to do performantly when using sidekiq with ActiveJob.
+
+When using Sidekiq via Active Job there are, confusingly, multiple levels of retries in the event of a job failure. Active Job has it's own retry mechanism, which once completed then passes down to Sidekiq’s own (seperate and different!) retry system. This is a strange thing to debug when things go wrong.
+
+When using Active Job and passing in an Active Record object, you rely on the object being serialized to a text argument using [Global ID](https://github.com/rails/globalid). This does save doing the lookup yourself (as you can see in the Active Job example) but it raises errors if records can be deleted (or even not created!) before the job is pulled form the queue. It also makes the arguments of the jobs a bit ugly in the web dashboard.
+
+There’s no real harm to “locking yourself into” a specific job framework. You are unlikely to swap in future for another queuing system with similar (intentionally) limited functionality availible in Active Job. In a real life situation where your asyncronous infrastructure is struggling, you’d be assessing various solutions with more functionality and planning for a major migration project.
+
+If you're using Sidekiq in prduction you should buy a license for the Pro version. The [extra reliability](https://github.com/mperham/sidekiq/wiki/Reliability#using-super_fetch) so that jobs are not “lost” due to crashes is enough of a reason when in a production environment, but there are other additional features including job deletion (by job id and kind of job) and web dashboard enhancements.
+
+
+## Why not?
+
+When using Sidekiq directly you have to think more about your job arguments and perform object lookups yourself, Sidekiq only permits simple values as job arguments.
+
+You can use Active Job-style _and_ direct Sidekiq jobs onto the same queues at the same time because Active Job wraps Sidekiq. You don't have to switch everything, you can use both styles in the same application.
+
+There is the deeper question as to why use Sidekiq at all if your workload is light, or you are early on in your project.
+
+[Good Job](https://github.com/bensheldon/good_job) and [Que](https://github.com/que-rb/que) (postgres only) or [Delayed Job](https://github.com/collectiveidea/delayed_job) (any SQL database) are well-regarded Active Job adapters and all use your existing database (in better or worse ways depending or your specific application) and then you don’t need to run an additional piece of database infrastructure.
```

Comment:
After the parentheses, start a new sentence: If you use one of these, you then don't need to run...